File: samples/precompiled/BackupSiteContent.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

/* This sample demonstrates the Fan-out/Fan-in pattern. In this pattern, multiple activity
 * functions are executed in parallel, and then the orchestrator waits for all of them to
 * complete before continuing. This is useful when you need to perform the same operation
 * on multiple items concurrently, such as backing up multiple files.
 *
 * Pattern documentation:
 * https://docs.microsoft.com/azure/azure-functions/durable/durable-functions-cloud-backup
 *
 * To run this sample:
 *   1. Configure the AzureWebJobsStorage connection string in local.settings.json
 *   2. Start the function app locally using `func host start` or run from Visual Studio
 *   3. Make an HTTP POST request to: http://localhost:7071/orchestrators/E2_BackupSiteContent
 *      Optionally include a JSON body with the root directory path to backup
 *
 * Required app settings:
 *   - AzureWebJobsStorage: Azure Storage connection string for storing backup blobs
 */
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.DurableTask;
using Microsoft.Extensions.Logging;
using Microsoft.WindowsAzure.Storage.Blob;

namespace VSSample
{
    public static class BackupSiteContent
    {
        // Orchestrator function that demonstrates Fan-out/Fan-in by backing up files in parallel.
        // First gets the file list (single activity), then fans out to copy each file concurrently.
        [FunctionName("E2_BackupSiteContent")]
        public static async Task<long> Run(
            [OrchestrationTrigger] IDurableOrchestrationContext backupContext)
        {
            // Get the root directory from input, or default to the assembly directory
            string rootDirectory = backupContext.GetInput<string>()?.Trim();
            if (string.IsNullOrEmpty(rootDirectory))
            {
                rootDirectory = Directory.GetParent(typeof(BackupSiteContent).Assembly.Location).FullName;
            }

            // Step 1: Get the list of files to backup (single activity call)
            string[] files = await backupContext.CallActivityAsync<string[]>(
                "E2_GetFileList",
                rootDirectory);

            // Step 2: Fan-out - start all file copy tasks in parallel (no await yet)
            var tasks = new Task<long>[files.Length];
            for (int i = 0; i < files.Length; i++)
            {
                tasks[i] = backupContext.CallActivityAsync<long>(
                    "E2_CopyFileToBlob",
                    files[i]);
            }

            // Step 3: Fan-in - wait for all parallel tasks to complete
            await Task.WhenAll(tasks);

            // Aggregate results from all tasks
            long totalBytes = tasks.Sum(t => t.Result);
            return totalBytes;
        }

        // Activity function that retrieves the list of files from a directory.
        [FunctionName("E2_GetFileList")]
        public static string[] GetFileList(
            [ActivityTrigger] string rootDirectory, 
            ILogger log)
        {
            log.LogInformation($"Searching for files under '{rootDirectory}'...");
            string[] files = Directory.GetFiles(rootDirectory, "*", SearchOption.AllDirectories);
            log.LogInformation($"Found {files.Length} file(s) under {rootDirectory}.");

            return files;
        }

        // Activity function that copies a single file to Azure Blob Storage.
        // This function is called in parallel for each file (fan-out).
        [FunctionName("E2_CopyFileToBlob")]
        public static async Task<long> CopyFileToBlob(
            [ActivityTrigger] string filePath,
            Binder binder,
            ILogger log)
        {
            long byteCount = new FileInfo(filePath).Length;

            // strip the drive letter prefix and convert to forward slashes
            string blobPath = filePath
                .Substring(Path.GetPathRoot(filePath).Length)
                .Replace('\\', '/');
            string outputLocation = $"backups/{blobPath}";

            log.LogInformation($"Copying '{filePath}' to '{outputLocation}'. Total bytes = {byteCount}.");

            // copy the file contents into a blob
            using (Stream source = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read))
            using (Stream destination = await binder.BindAsync<CloudBlobStream>(
                new BlobAttribute(outputLocation, FileAccess.Write)))
            {
                await source.CopyToAsync(destination);
            }

            return byteCount;
        }
    }
}

File: samples/fsharp/BackupSiteContent.fs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

namespace VSSample

open System
open System.IO
open System.Threading.Tasks
open Microsoft.Azure.WebJobs
open Microsoft.Azure.WebJobs.Extensions.DurableTask
open Microsoft.Extensions.Logging
open Microsoft.WindowsAzure.Storage.Blob
open FSharp.Control.Tasks

module BackupSiteContent =

  [<FunctionName("E2_BackupSiteContent")>]
  let Run([<OrchestrationTrigger>] backupContext: IDurableOrchestrationContext) = task {
    let input = backupContext.GetInput<string>()
    let rootDirectory = 
      if String.IsNullOrEmpty(input) then input.Trim()
      else Directory.GetParent(System.Reflection.Assembly.GetExecutingAssembly().Location).FullName

    let! files = backupContext.CallActivityAsync<string[]>("E2_GetFileList", rootDirectory)
    let tasks = files |> Array.map (fun f -> backupContext.CallActivityAsync<int64>("E2_CopyFileToBlob", f))
    let! results = Task.WhenAll tasks

    let totalBytes = Array.sum results
    return totalBytes
  }
  
  [<FunctionName("E2_GetFileList")>]
  let GetFileList([<ActivityTrigger>] rootDirectory: string, log: ILogger) =
    log.LogInformation (sprintf "Searching for files under '%s'..." rootDirectory)
    let files = Directory.GetFiles(rootDirectory, "*", SearchOption.AllDirectories)
    log.LogInformation (sprintf "Found %i file(s) under %s." files.Length rootDirectory)
    files

  [<FunctionName("E2_CopyFileToBlob")>]
  let CopyFileToBlob([<ActivityTrigger>] filePath: string, binder: Binder, log: ILogger) = task {
    let byteCount = FileInfo(filePath).Length

    // strip the drive letter prefix and convert to forward slashes
    let blobPath = filePath.Substring(Path.GetPathRoot(filePath).Length).Replace('\\', '/')
    let outputLocation = "backups/" + blobPath

    log.LogInformation (sprintf "Copying '%s' to '%s'. Total bytes = %i." filePath outputLocation byteCount)

    // copy the file contents into a blob
    use source = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read)
    let attribute = BlobAttribute(outputLocation, FileAccess.Write)
    use! destination = binder.BindAsync<CloudBlobStream> attribute
    do! source.CopyToAsync destination

    return byteCount
  }